python/ambassador/ir/irutils.py

```diff
@@ -16,33 +16,139 @@
+
+    match = g2match.endswith(g1match)
+    # logging.debug("  match start: %s ~ %s => %s", g1match, g2match, match)
+
+    return match
+
+
+# hostglob_matches_end has g1 ending with '*' and g2 not starting with '*';
+# it's OK for g2 to end with a wilcard too.
+
+def hostglob_matches_end(g1, g2, g2end):
+    # Leading "*" cannot match an empty string, so unless we have a wildcard
+    # for g2, we have to have g1 longer than g2.
+    g1match = g1[:-1]
+    g2match = g2[:-1] if g2end else g2
+
+    if len(g1) > len(g2match):
+        if not g2end:
+            # logging.debug("  match end: %s is too short => False", g1)
+            return False
+
+        # Wildcards for both, so make sure we do the substring match against
+        # the longer one.
+        g1match = g2[:-1]
+        g2match = g1[:-1]
+
+    match = g2match.startswith(g1match)
+    # logging.debug("  match end: %s ~ %s => %s", g1match, g2match, match)
+
+    return match
+
+
 ################
-## hostglob_matches is a utility for host globbing.
-
-def hostglob_matches(glob: str, value: str) -> bool:
-    """
-    Does a host glob match a given value?
-    """
-
-    rc = False
-
-    if ('*' in value) and not ('*' in glob):
-        # Swap.
-        tmp = value
-        value = glob
-        glob = tmp
-
-    if glob == "*": # special wildcard
-        rc=True
-    elif glob.endswith("*"): # prefix match
-        rc=value.startswith(glob[:-1])
-    elif glob.startswith("*"): # suffix match
-        rc=value.endswith(glob[1:])
-    else: # exact match
-        rc=value == glob
-
-    # sys.stderr.write(f"hostglob_matches: {value} gl~ {glob} == {rc}\n")
-    return rc
+## hostglob_matches determines two given DNS globs fit each other,
+## in either direction.
+
+def hostglob_matches(g1: str, g2: str) -> bool:
+    # logging.debug("hostglob_matches: %s ~ %s", g1, g2)
+
+    # Short-circuit: if g1 & g2 are equal, we're done here.
+    if g1 == g2:
+        # logging.debug("  equal => True")
+        return True
+
+    # Next special case: if either glob is "*", then it matches everything.
+    if (g1 == "*") or (g2 == "*"):
+        # logging.debug("  \"*\" present => True")
+        return True
+
+    # Final special case: if either starts with a bare ".", that's not OK.
+    # (Ending with a bare "." is different because DNS.)
+    if g1[0] == "." or g2[0] == ".":
+        # logging.debug("  exact match starts with bare \".\" => False")
+        return False
+
+    # OK, we don't have the simple-"*" case, so any wildcards must be at
+    # the start or end, and they must be a component alone.
+    g1start = (g1[0]  == "*")
+    g1end =   (g1[-1] == "*")
+    g2start = (g2[0]  == "*")
+    g2end =   (g2[-1] == "*")
+
+    # logging.debug("  g1start=%s g1end=%s g2start=%s g2end=%s", g1start, g1end, g2start, g2end)
+
+    if (g1start and g1end) or (g2start and g2end):
+        # Not a valid DNS glob: you can't have a "*" at both ends.
+        return g1 == g2
```

Comment:
FWIW, if this happens Envoy would treat the leading `*` as a wildcard, and the trailing `*` as a literal.